frontend/js/src/add-listen/add-listen-modal.tsx

```diff
@@ -0,0 +1,396 @@
+  "track",
+  "album",
+}
+
+export interface AddListenModalProps {
+  newAlert: (
+    alertType: AlertType,
+    title: string,
+    message: string | JSX.Element
+  ) => void;
+}
+
+export interface AddListenModalState {
+  listenOption: SubmitListenType;
+  selectedTrack?: ACRMSearchResult;
+  customTimestamp: Boolean;
+  selectedDate: Date;
+  searchField: string;
+  trackResults: Array<ACRMSearchResult>;
+}
+
+export default class AddListenModal extends React.Component<
+  AddListenModalProps,
+  AddListenModalState
+> {
+  static contextType = GlobalAppContext;
+  declare context: React.ContextType<typeof GlobalAppContext>;
+
+  constructor(props: AddListenModalProps) {
+    super(props);
+    this.state = {
+      listenOption: SubmitListenType.track,
+      selectedTrack: undefined,
+      customTimestamp: false,
+      selectedDate: new Date(),
+      searchField: "",
+      trackResults: [],
+    };
+  }
+
+  static getListenFromTrack = (
+    selectedDate: Date,
+    selectedTrack?: ACRMSearchResult
+  ): Listen | undefined => {
+    if (!selectedTrack) {
+      return undefined;
+    }
+
+    return {
+      listened_at: convertDateToUnixTimestamp(selectedDate),
+      track_metadata: {
+        additional_info: {
+          release_mbid: selectedTrack.release_mbid,
+          recording_mbid: selectedTrack.recording_mbid,
+        },
+
+        artist_name: selectedTrack.artist_credit_name,
+        track_name: selectedTrack.recording_name,
+        release_name: selectedTrack.release_name,
+      },
+    };
+  };
+
+  handleError = (error: string | Error, title?: string): void => {
+    const { newAlert } = this.props;
+    if (!error) {
+      return;
+    }
+    newAlert(
+      "danger",
+      title || "Error",
+      typeof error === "object" ? error.message : error
+    );
+  };
+
+  trackMetadata = (track: ACRMSearchResult) => {
+    this.setState({
+      selectedTrack: track,
+    });
+  };
+
+  submitListen = async () => {
+    const { newAlert } = this.props;
+    const { APIService, currentUser } = this.context;
+    const { selectedTrack, selectedDate } = this.state;
+    if (currentUser?.auth_token) {
+      if (selectedTrack) {
+        const payload = AddListenModal.getListenFromTrack(
+          selectedDate,
+          selectedTrack
+        );
+
+        if (currentUser.auth_token !== undefined) {
+          try {
+            if(payload){
+            const status = await APIService.submitListens(
+              currentUser.auth_token,
+              "single",
+              [payload]
+            );
+            if (status.status === 200) {
+              newAlert(
+                "success",
+                "You added the listen",
+                `${selectedTrack.recording_name} - ${selectedTrack.artist_credit_name}`
+              );
+              this.resetTrackSelection();
+            }
+          }} catch (error) {
+            this.handleError(error, "Error while adding a listen");
+          }
```

Comment:
This is not currently catching failed requests.
It will catch a network error on the client side (that will throw an error, caught in your catch block underneath) but if the request is successful but returns an error code it will fail silently. That's one of the gotchas of the fetch API; error management is complex.

I tried locally with an invalid token and got a 401 response, saw there was no error shown on screen.

Here's what I suggest, with a description of my changes below:
```typescript
submitListen = async () => {
    const { newAlert } = this.props;
    const { APIService, currentUser } = this.context;
    const { selectedTrack, selectedDate } = this.state;
    if (currentUser?.auth_token) {
      if (selectedTrack) {
        const payload = AddListenModal.getListenFromTrack(
          selectedDate,
          selectedTrack
        )!;
        try {
          const response = await APIService.submitListens(
            currentUser.auth_token,
            "single",
            [payload]
          );
          await APIService.checkStatus(response);
          newAlert(
            "success",
            "You added the listen",
            `${selectedTrack.recording_name} - ${selectedTrack.artist_credit_name}`
          );
          this.resetTrackSelection();
        } catch (error) {
          this.handleError(error, "Error while adding a listen");
        }
      }
    } else {
      newAlert(
        "danger",
        "You need to be logged in to Add a Listen",
        <a href="/login">Log in here</a>
      );
    }
  };
```
Changes:
1. Where we create the `payload`, I added a `!` at the end (easy to miss) which is a clue to tell typescript "this value can't be null or undefined". Since we have a check for `selectedTrack` just above we can guarantee payload won't be undefined, allowing me to removed the `if(payload)` check.
2. Changed the variable name from `status` to `response` since that's what `APIService.submitListens` returns. Naming is important! Reading `status.status` made me twitch :D
3. I used the utility function `checkStatus` we use to handle fetch requests. It will do nothing if the request is successful and code execution will proceed. If the request is not successful it will parse the error message and `throw` an error. Sicne you have a catch block underneath, that's what we want.
4. Tadaaa: 

![image](https://user-images.githubusercontent.com/6179856/221849975-834ac486-eedf-474d-bd1b-0428f081bbe2.png)